libjwt/jwt-openssl.c

```diff
@@ -456,3 +472,136 @@ int jwt_verify_sha_pem(jwt_t *jwt, const char *head, const char *sig_b64)
+    json_t *jwks = NULL;
+    const char* kid = NULL;
+    size_t index = 0;
+    json_t *value = NULL;
+
+    js = json_loads(sig_jwks, 0, NULL);
+    
+    if (json_is_object(js))
+    {
+        jwk = json_object_get(js, "key");
+        jwks = json_object_get(js, "keys");
+        
+        if (json_is_array(jwks) && jwt->kid) {
+            // Find the jwk that matches the given key id (kid)
+            json_array_foreach(jwks, index, value) {
+                jwk = json_array_get( jwks, index );
+                kid = json_string_value(json_object_get(value, "kid"));
+                if(strcmp(kid, jwt->kid)==0) {
+                    jwk = value;
+                    break;
+                }
+            }
+        }
+        
+        if (json_is_object(jwk)) {
+            // Retain the found JWK
+            json_incref(jwk);
+        }
+        
+        // Release the parsed JSON
+        json_decref(js);
+    }
+    
+    return jwk;
+}
+
+EVP_PKEY *jwt_create_key_from_jwk(json_t *jwk)
+{
+    const char *kty = NULL;
+    kty = json_string_value(json_object_get(jwk, "kty"));
+    
+    if (strcmp(kty,"RSA")==0)
+        return jwt_create_rsa_from_jwk(jwk);
+    else if( strcmp(kty,"EC")==0)
+        return jwt_create_ec_from_jwk(jwk);
+    else
+        return NULL;
+}
+
+EVP_PKEY *jwt_create_rsa_from_jwk(json_t *jwk)
+{
+    RSA *rsa = NULL;
+    BIGNUM *n = NULL;
+    BIGNUM *e = NULL;
+    BIGNUM *p = NULL;
+    BIGNUM *q = NULL;
+    BIGNUM *dp = NULL;
+    BIGNUM *dq = NULL;
+    BIGNUM *qi = NULL;
+    EVP_PKEY *pkey = NULL;
+    
+    n = jwt_create_big_num(jwk, "n");
+    e = jwt_create_big_num(jwk, "e");
+    p = jwt_create_big_num(jwk, "p");
+    q = jwt_create_big_num(jwk, "q");
+    dp = jwt_create_big_num(jwk, "dp");
+    dq = jwt_create_big_num(jwk, "dq");
+    qi = jwt_create_big_num(jwk, "qi");
+    
+    rsa = RSA_new();
+    
+    if (n && e) {
+        rsa->n = n;
+        rsa->e = e;
+        rsa->d = NULL;
+    }
+    
+    if (p && q ) {
+        rsa->p = p;
+        rsa->q = q;
+    }
+    
+    if (dp && dq && qi ) {
+        rsa->dmp1 = dp;
+        rsa->dmq1 = dq;
+        rsa->dmq1 = qi;
+    }
+    
+    RSA_up_ref(rsa);
+    
+    pkey = EVP_PKEY_new();
+    if (EVP_PKEY_set1_RSA(pkey, rsa)==0) {
+        EVP_PKEY_free(pkey);
+        pkey = NULL;
+    }
+    
+    return pkey;
+}
+
+BIGNUM *jwt_create_big_num( json_t *jwk, const char *property )
+{
+    const char *b64 = NULL;
+    void *data = NULL;
+    int size = 0;
+    BIGNUM *bigNum = NULL;
+    
+    // Read the base64 encoded string from the jwk
+    b64 = json_string_value(json_object_get(jwk, property));
+    
+    // Decode the base64 encoded string into binary form
+    if (b64)
+        data = jwt_b64_decode( b64, &size );
+    
+    // Create a BIGNUM from the binary data
+    if (data) {
+        bigNum = BN_bin2bn(data,size,NULL);
+        free(data);
+    }
+    
+    return bigNum;
+}
+
+EVP_PKEY *jwt_create_ec_from_jwk(json_t *jwk)
```

Comment:
Let's keep stubs out of the proper code. Thanks.

Reply:
I have removed the stub.